src/Sylius/Bundle/ApiBundle/Controller/TokenController.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ApiBundle\Controller;
+
+use FOS\RestBundle\Decoder\DecoderProviderInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\ParameterBag;
+use OAuth2\OAuth2;
+use OAuth2\OAuth2ServerException;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+/**
+ * @see \FOS\OAuthServerBundle\Controller\TokenController
+ * @see \FOS\RestBundle\EventListener\BodyListener
+ *
+ * @author Łukasz Chruściel <lchrusciel@gmail.com>
+ */
+final class TokenController
+{
+    /**
+     * @var OAuth2
+     */
+    private $server;
+
+    /**
+     * @var DecoderProviderInterface
+     */
+    private $decoderProvider;
+
+    /**
+     * @param OAuth2 $server
+     * @param DecoderProviderInterface $decoderProvider
+     */
+    public function __construct(OAuth2 $server, DecoderProviderInterface $decoderProvider)
+    {
+        $this->server = $server;
+        $this->decoderProvider = $decoderProvider;
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function tokenAction(Request $request)
+    {
+        $this->convertContentToRequestParameters($request);
+
+        try {
+            return $this->server->grantAccessToken($request);
+        } catch (OAuth2ServerException $e) {
+            return $e->getHttpResponse();
+        }
+    }
+
+    /**
+     * Additional action to keep default behaviour of FosRestBundle. Main aim is to turn off data normalization.
+     *
+     * @param Request $request
+     */
+    private function convertContentToRequestParameters(Request $request)
+    {
+        $contentType = $request->headers->get('Content-Type');
+        $format = is_null($contentType) ? $request->getRequestFormat() : $request->getFormat($contentType);
+        $content = $request->getContent();
+
+        if (!$this->decoderProvider->supports($format) && empty($content)) {
```

Comment:
Can we extract the second condition to another `if` before this one? Are they really related? Should we decode every supported format even if no content was sent?